ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -138,91 +281,140 @@ public StatsSnapshot getStatsSnapshot() throws StoreException {
           new AtomicLong(0));
     } catch (IOException e) {
-      throw new StoreException("I/O error while getting entries from index segment", e, StoreErrorCodes.IOError);
+      throw new StoreException("I/O exception while getting entries from index segment", e, StoreErrorCodes.IOError);
     }
+    return indexEntries;
+  }
+
+  /**
+   * Get a {@link List} of valid {@link IndexValue}s from a given {@link List} of {@link IndexEntry} that belong to the
+   * same {@link IndexSegment}.
+   * @param indexSegment the {@link IndexSegment} where the entries came from
+   * @param indexEntries a {@link List} of unfiltered {@link IndexEntry} whose elements could be valid or invalid
+   * @param deleteAndExpirationRefTimeInMs the reference time in ms until which deletes and expiration are relevant
+   * @param deletedKeys a {@link Map} of deleted keys to operation time. Used to determine whether a PUT is deleted
+   * @return a {@link List} of valid {@link IndexValue}
+   * @throws StoreException
+   */
+  private List<IndexEntry> getValidIndexEntries(IndexSegment indexSegment, List<IndexEntry> indexEntries,
+      long deleteAndExpirationRefTimeInMs, Map<StoreKey, Long> deletedKeys) throws StoreException {
+    List<IndexEntry> validIndexEntries = new ArrayList<>();
     for (IndexEntry indexEntry : indexEntries) {
       IndexValue indexValue = indexEntry.getValue();
       if (indexValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-        // delete record
-        validIndexValues.add(indexValue);
+        // delete record is always valid
+        validIndexEntries.add(indexEntry);
+        if (isExpired(indexValue.getExpiresAtMs(), deleteAndExpirationRefTimeInMs)) {
+          continue;
+        }
         long operationTimeInMs =
             indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
                 : indexValue.getOperationTimeInMs();
-        if (operationTimeInMs < deleteReferenceTimeInMs) {
-          // delete is relevant
-          deletedKeys.add(indexEntry.getKey());
-        } else if (!isExpired(indexValue.getExpiresAtMs(), expirationReferenceTimeInMs)
-            && indexValue.getOriginalMessageOffset() != -1
-            && indexValue.getOriginalMessageOffset() >= indexSegment.getStartOffset().getOffset()) {
-          // delete is irrelevant but it's in the same index segment as the put and the put is still valid
-          BlobReadOptions originalPut =
-              index.getBlobReadInfo(indexEntry.getKey(), EnumSet.of(StoreGetOptions.Store_Include_Deleted));
-          Offset originalPutOffset = new Offset(originalPut.getLogSegmentName(), originalPut.getOffset());
-          validIndexValues.add(new IndexValue(originalPut.getSize(), originalPutOffset, originalPut.getExpiresAtMs(),
-              Utils.Infinite_Time, indexValue.getServiceId(), indexValue.getContainerId()));
+        deletedKeys.put(indexEntry.getKey(), operationTimeInMs);
```

Comment:
# of delete record * (size of StoreKey + 8 bytes (long))
if we assume 1.9m record per store, each entry is 100 bytes and all records are delete record then roughly we need 190MB per store.